src/network/networkd-mdb.c

```diff
@@ -0,0 +1,245 @@
+        if (r < 0)
+                return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                  "Cannot parse group address: %m");
+
+        if (!in_addr_is_multicast(mdb_entry->family, &mdb_entry->group_addr))
+                return log_syntax(unit, LOG_WARNING, filename, line, -EINVAL,
+                                  "Address %s is not a multicast group", rvalue);
+
+        mdb_entry = NULL;
+
+        return 0;
+}
+
+/* remove and MDB entry. */
+void mdb_entry_free(MdbEntry *mdb_entry) {
+        if (!mdb_entry)
+                return;
+
+        if (mdb_entry->network) {
+                LIST_REMOVE(static_mdb_entries, mdb_entry->network->static_mdb_entries, mdb_entry);
+                assert(mdb_entry->network->n_static_mdb_entries > 0);
+                mdb_entry->network->n_static_mdb_entries--;
+
+                if (mdb_entry->section)
+                        hashmap_remove(mdb_entry->network->mdb_entries_by_section, mdb_entry->section);
+        }
+
+        network_config_section_free(mdb_entry->section);
+        free(mdb_entry);
+}
+
+static int set_mdb_handler(sd_netlink *rtnl, sd_netlink_message *m, Link *link) {
+        int r;
+
+        assert(link);
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 1;
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0 && r != -EEXIST) {
+                log_link_message_warning_errno(link, m, r, "Could not add MDB entry");
+                link_enter_failed(link);
+                return 1;
+        }
+
+        return 1;
+}
+
+/* send a request to the kernel to add an MDB entry */
+int mdb_entry_configure(Link *link, MdbEntry *mdb_entry) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *req = NULL;
+        struct br_mdb_entry entry = {};
+        int r;
+
+        assert(link);
+        assert(link->network);
+        assert(link->network->bridge);
+        assert(link->manager);
+        assert(mdb_entry);
+
+        entry.state = MDB_PERMANENT;
+        entry.ifindex = link->ifindex;
+        entry.vid = mdb_entry->vlan_id;
+
+        /* create new RTM message */
+        r = sd_rtnl_message_new_mdb(link->manager->rtnl, &req, RTM_NEWMDB, link->network->bridge->ifindex);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not create RTM_NEWMDB message: %m");
+
+        switch (mdb_entry->family) {
+        case AF_INET:
+                entry.addr.u.ip4 = mdb_entry->group_addr.in.s_addr;
+                entry.addr.proto = htons(ETH_P_IP);
+                break;
+
+        case AF_INET6:
+                memcpy(&entry.addr.u.ip6, &mdb_entry->group_addr.in6, sizeof(entry.addr.u.ip6));
+                entry.addr.proto = htons(ETH_P_IPV6);
+                break;
+
+        default:
+                return log_link_error_errno(link, -EINVAL, "Invalid network family");
```

Comment:
log functions should only take error number we *receive* as-is. If we make up our own error numbers please use SYNTHETIC_ERRNO(). Also, I think EAFNOSUPPORT might be slightly better an error code here.

```
return log_link_error_errno(link, SYNTHETIC_ERRNO(EAFNOSUPPORT), "…
```